java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java

```diff
@@ -0,0 +1,192 @@
+public class EqualsArgumentTypeCheck extends SubscriptionBaseVisitor {
+
+  private static final MethodInvocationMatcher EQUALS_MATCHER = MethodInvocationMatcher.create()
+    .name("equals")
+    .addParameter("java.lang.Object");
+
+  private boolean typeChecked;
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (hasSemantic()) {
+      MethodTree methodTree = (MethodTree) tree;
+      if (methodTree.block() != null && "equals".equals(methodTree.symbol().name()) && methodTree.parameters().size() == 1) {
+        Symbol parameterSymbol = methodTree.parameters().get(0).symbol();
+        if (parameterSymbol.type().is("java.lang.Object")) {
+          typeChecked = false;
+          methodTree.accept(new EqualsMethodVisitor(parameterSymbol));
+          if (!typeChecked) {
+            addIssue(tree, "Add a type test to this method.");
+          }
+        }
+      }
+    }
+  }
+
+  private class EqualsMethodVisitor extends BaseTreeVisitor {
+    private final Symbol parameterSymbol;
+
+    EqualsMethodVisitor(Symbol parameterSymbol) {
+      this.parameterSymbol = parameterSymbol;
+    }
+
+    @Override
+    public void visitExpressionStatement(ExpressionStatementTree tree) {
+      tree.expression().accept(new ExpressionVisitor(parameterSymbol));
+    }
+
+    @Override
+    public void visitIfStatement(IfStatementTree tree) {
+      tree.condition().accept(new ExpressionVisitor(parameterSymbol));
+      super.visitIfStatement(tree);
+    }
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree tree) {
+      if (tree.expression() != null) {
+        tree.expression().accept(new ExpressionVisitor(parameterSymbol));
+      }
+    }
+
+    @Override
+    public void visitVariable(VariableTree tree) {
+      if (tree.initializer() != null) {
+        tree.initializer().accept(new ExpressionVisitor(parameterSymbol));
+      }
+      super.visitVariable(tree);
+    }
+  }
+
+  private class ExpressionVisitor extends BaseTreeVisitor {
+    private final Symbol parameterSymbol;
+
+    ExpressionVisitor(Symbol parameterSymbol) {
+      this.parameterSymbol = parameterSymbol;
+    }
+
+    @Override
+    public void visitInstanceOf(InstanceOfTree tree) {
+      if (isArgument(tree.expression())) {
+        typeChecked = true;
+      }
+    }
+
+    @Override
+    public void visitBinaryExpression(BinaryExpressionTree tree) {
+      if (tree.is(Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO)) {
+        if (isGetClassOnArgument(tree.leftOperand()) || isGetClassOnArgument(tree.rightOperand())) {
+          typeChecked = true;
+        } else if (isExplicitComparison(tree.leftOperand(), tree.rightOperand())) {
+          typeChecked = true;
+        }
+      } else {
+        super.visitBinaryExpression(tree);
+      }
+    }
+
+    public void visitMethodInvocation(MethodInvocationTree tree) {
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Add the "@Override" annotation above this method signature[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)